packages/server/graphql/mutations/helpers/notifications/MSTeamsNotifier.ts

```diff
@@ -0,0 +1,419 @@
+    return notifyMSTeams('meetingEnd', webhookUrl, facilitatorUserId, team.id, attachments)
+  },
+
+  async startTimeLimit(scheduledEndTime, meeting, team) {
+    const {webhookUrl} = notificationChannel
+
+    const {phases, facilitatorStageId, facilitatorUserId} = meeting
+
+    const maybeMeetingShortLink = makeAppURL(process.env.INVITATION_SHORTLINK!, `${meeting.id}`)
+    const meetingUrl = makeAppURL(appOrigin, `meet/${meeting.id}`)
+
+    const stageRes = findStageById(phases, facilitatorStageId)
+    const {stage} = stageRes!
+    const {phaseType} = stage
+    const phaseLabel = phaseLabelLookup[phaseType]
+
+    const meetingTitle = `The **${phaseLabel} Phase** has begun ⌛`
+
+    const card = new AdaptiveCards.AdaptiveCard()
+    card.version = new AdaptiveCards.Version(1.2, 0)
+
+    const titleTextBlock = GenerateACMeetingTitle(meetingTitle)
+    card.addItem(titleTextBlock)
+
+    const meetingDetailColumnSet = GenerateACMeetingAndTeamsDetails(team, meeting)
+    card.addItem(meetingDetailColumnSet)
+
+    const teamFixedTime = scheduledEndTime.toISOString().replace(/.\d+Z$/g, 'Z')
+    const timeLimitText = `You have until {{DATE(${teamFixedTime},SHORT)}} at {{TIME(${teamFixedTime})}} to complete it.`
+
+    const timeLimitColumnSet = new AdaptiveCards.ColumnSet()
+    timeLimitColumnSet.spacing = AdaptiveCards.Spacing.ExtraLarge
+    const timeLimitColumn = new AdaptiveCards.Column()
+    timeLimitColumn.width = 'stretch'
+    const timeLimitTextBlock = new AdaptiveCards.TextBlock(timeLimitText)
+    timeLimitTextBlock.wrap = true
+
+    timeLimitColumn.addItem(timeLimitTextBlock)
+    timeLimitColumnSet.addColumn(timeLimitColumn)
+
+    card.addItem(timeLimitColumnSet)
+
+    const meetingLinkColumnSet = new AdaptiveCards.ColumnSet()
+    meetingLinkColumnSet.spacing = AdaptiveCards.Spacing.ExtraLarge
+    const meetingLinkColumn = new AdaptiveCards.Column()
+    meetingLinkColumn.width = 'stretch'
+    const meetingLinkHeaderTextBlock = new AdaptiveCards.TextBlock('Link: ')
+    meetingLinkHeaderTextBlock.wrap = true
+    meetingLinkHeaderTextBlock.weight = AdaptiveCards.TextWeight.Bolder
+    const meetingLinkTextBlock = new AdaptiveCards.TextBlock(maybeMeetingShortLink)
+    meetingLinkTextBlock.color = AdaptiveCards.TextColor.Accent
+    meetingLinkTextBlock.size = AdaptiveCards.TextSize.Small
+    meetingLinkTextBlock.wrap = true
+    const joinMeetingActionSet = new AdaptiveCards.ActionSet()
+    const joinMeetingAction = new AdaptiveCards.OpenUrlAction()
+    joinMeetingAction.title = 'Open Meeting'
+    joinMeetingAction.url = meetingUrl
+    joinMeetingAction.id = 'openMeeting'
+
+    joinMeetingActionSet.addAction(joinMeetingAction)
+
+    meetingLinkColumn.addItem(meetingLinkHeaderTextBlock)
+    meetingLinkColumn.addItem(meetingLinkTextBlock)
+    meetingLinkColumn.addItem(joinMeetingActionSet)
+
+    meetingLinkColumnSet.addColumn(meetingLinkColumn)
+
+    card.addItem(meetingLinkColumnSet)
```

Comment:
+1 This file is still 400+ lines long, making it difficult for the next developer to read and make edits. I think it could still be significantly smaller if we use more helper functions which could exist in separate files. 

For example, many of the lines in each notification type are repeated several times. The main differences are the title and meetingLinkHeaderTextBlock, which could be passed in as variables.